File: tape/sdk/go/client_test.go
package tape

import (
	"context"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"testing"
	"time"

	pb "github.com/vamsiramakrishnan/durable-agents/tape/sdk/go/tapepb"
)

func freePort(t *testing.T) int {
	t.Helper()
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	p := l.Addr().(*net.TCPAddr).Port
	l.Close()
	return p
}

func waitFor(t *testing.T, addr string, timeout time.Duration) {
	t.Helper()
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		c, err := net.DialTimeout("tcp", addr, time.Second)
		if err == nil {
			c.Close()
			return
		}
		time.Sleep(100 * time.Millisecond)
	}
	t.Fatalf("server never came up at %s", addr)
}

func startServer(t *testing.T) (string, func()) {
	t.Helper()
	bin, err := filepath.Abs("../../server/target/debug/tape-server")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(bin); err != nil {
		t.Skipf("tape-server not built: %v (run `cargo build` in tape/server)", err)
	}
	port := freePort(t)
	addr := "127.0.0.1:" + itoa(port)
	cmd := exec.Command(bin, "--listen", addr, "--store", "memory")
	cmd.Stderr = nil
	cmd.Stdout = nil
	cmd.Env = append(os.Environ(), "RUST_LOG=tape_server=warn")
	if err := cmd.Start(); err != nil {
		t.Fatal(err)
	}
	waitFor(t, addr, 15*time.Second)
	return "tape://" + addr, func() {
		_ = cmd.Process.Kill()
		_, _ = cmd.Process.Wait()
	}
}

func itoa(n int) string {
	const d = "0123456789"
	if n == 0 {
		return "0"
	}
	var s []byte
	for n > 0 {
		s = append([]byte{d[n%10]}, s...)
		n /= 10
	}
	return string(s)
}

func TestClientRoundTripsLifecycle(t *testing.T) {
	url, stop := startServer(t)
	defer stop()
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	c, err := Dial(url)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	r, err := c.BeginRun(ctx, BeginRunOpts{
		AppName: "a", UserID: "u", SessionID: "go-smoke",
		InvocationID: "inv-go", LeaseOwner: "test", LeaseTTLMs: 60_000,
	})
	if err != nil {
		t.Fatal(err)
	}
	if r.Resumed {
		t.Fatalf("expected fresh run; got resumed=%v", r.Resumed)
	}

	if _, err := c.RecordDecision(ctx, r.RunId, 0, "m", "{}", `{"plan":1}`, "", "p1"); err != nil {
		t.Fatal(err)
	}
	if d, err := c.GetDecision(ctx, r.RunId, 0); err != nil || !d.Found {
		t.Fatalf("get_decision: %v found=%v", err, d.Found)
	}

	be, err := c.BeginEffect(ctx, BeginEffectOpts{
		RunID: r.RunId, DecisionIndex: 0, ToolName: "execute_sweep",
		CallIndex: 0, RequestJSON: "{}",
	})
	if err != nil {
		t.Fatal(err)
	}
	if int32(be.Status) != EffectStatusPending {
		t.Fatalf("status=%v want PENDING", be.Status)
	}
	if be.IdempotencyKey != r.RunId+"/decision-0/execute_sweep/0" {
		t.Fatalf("key=%q", be.IdempotencyKey)
	}

	// short-circuit on second begin
	be2, err := c.BeginEffect(ctx, BeginEffectOpts{RunID: r.RunId, DecisionIndex: 0, ToolName: "execute_sweep", CallIndex: 0})
	if err != nil {
		t.Fatal(err)
	}
	if be2.IdempotencyKey != be.IdempotencyKey {
		t.Fatalf("short-circuit key changed")
	}

	if _, err := c.CompleteEffect(ctx, r.RunId, be.IdempotencyKey, EffectStatusConfirmed, `{"wire_id":"w1"}`, ""); err != nil {
		t.Fatal(err)
	}
	ge, err := c.GetEffect(ctx, r.RunId, be.IdempotencyKey)
	if err != nil || !ge.Found {
		t.Fatalf("get_effect: %v found=%v", err, ge.Found)
	}
	if int32(ge.Effect.Status) != EffectStatusConfirmed {
		t.Fatalf("status after complete: %v", ge.Effect.Status)
	}

	if _, err := c.RegisterCompensation(ctx, r.RunId, be.IdempotencyKey, "reverse_wire", "{}"); err != nil {
		t.Fatal(err)
	}
	obs, err := c.ListObligations(ctx, r.RunId, true)
	if err != nil || len(obs.Obligations) != 1 {
		t.Fatalf("obligations: %v len=%d", err, len(obs.Obligations))
	}

	if _, err := c.SetBudget(ctx, r.RunId, 1.0, 0); err != nil {
		t.Fatal(err)
	}
	adm, _ := c.AdmitBudget(ctx, r.RunId, 0.5, 0)
	if !adm.Admitted {
		t.Fatalf("expected admitted")
	}
	if _, err := c.ChargeBudget(ctx, r.RunId, 0.9, 0); err != nil {
		t.Fatal(err)
	}
	adm, _ = c.AdmitBudget(ctx, r.RunId, 0.5, 0)
	if adm.Admitted {
		t.Fatalf("expected refused after charging")
	}

	// timer
	tr, err := c.SetTimer(ctx, SetTimerOpts{
		RunID: r.RunId, FireAtMs: time.Now().Add(-time.Second).UnixMilli(),
		Kind: "gate_timeout", PayloadJSON: `{"gate":"g1"}`,
	})
	if err != nil {
		t.Fatal(err)
	}
	due, _ := c.ListDueTimers(ctx, 0, 50, true)
	found := false
	for _, x := range due.Timers {
		if x.TimerId == tr.TimerId {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("timer not in due list")
	}

	if _, err := c.EndRun(ctx, r.RunId, RunStatusTerminal, ""); err != nil {
		t.Fatal(err)
	}
	fresh, _ := c.GetRun(ctx, r.RunId)
	if int32(fresh.Status) != RunStatusTerminal {
		t.Fatalf("status: %v", fresh.Status)
	}

	// re-begin -> resumed=TERMINAL
	again, _ := c.BeginRun(ctx, BeginRunOpts{
		AppName: "a", UserID: "u", SessionID: "go-smoke",
		InvocationID: "inv-go", LeaseOwner: "test", LeaseTTLMs: 60_000,
	})
	if !again.Resumed || again.RunId != r.RunId || int32(again.Status) != RunStatusTerminal {
		t.Fatalf("re-begin: resumed=%v rid=%q status=%v", again.Resumed, again.RunId, again.Status)
	}

	// ── outbox / non-idempotent contract ───────────────────────────────────
	// A second run with a NON_IDEMPOTENT + OUTBOX effect, plus the four new
	// outbox/observation RPCs. Proves the Go SDK and the server agree on the
	// new wire.
	r2, err := c.BeginRun(ctx, BeginRunOpts{
		AppName: "a", UserID: "u", SessionID: "go-outbox",
		InvocationID: "inv-go-outbox", LeaseOwner: "test", LeaseTTLMs: 60_000,
	})
	if err != nil {
		t.Fatal(err)
	}
	// Server refuses NON_IDEMPOTENT + INLINE:
	if _, err := c.BeginEffect(ctx, BeginEffectOpts{
		RunID: r2.RunId, DecisionIndex: -1, ToolName: "wire_money",
		Semantics: EffectSemanticsNonIdempotent, DispatchMode: EffectDispatchInline,
	}); err == nil {
		t.Fatalf("expected NON_IDEMPOTENT+INLINE to be refused")
	}
	// NON_IDEMPOTENT + OUTBOX is accepted; business_key is enforced unique.
	oe, err := c.BeginEffect(ctx, BeginEffectOpts{
		RunID: r2.RunId, DecisionIndex: -1, ToolName: "wire_money",
		RequestJSON: `{"amount":100}`,
		Semantics: EffectSemanticsNonIdempotent, DispatchMode: EffectDispatchOutbox,
		BusinessKey: "go:bk-1", Connector: "bank.wire",
	})
	if err != nil {
		t.Fatal(err)
	}
	if int32(oe.Status) != EffectStatusPending {
		t.Fatalf("outbox effect status=%v", oe.Status)
	}
	// Listed by ListEffectsToDispatch.
	dispatchList, err := c.ListEffectsToDispatch(ctx, "bank.wire", 50)
	if err != nil {
		t.Fatal(err)
	}
	gotIt := false
	for _, e := range dispatchList.Effects {
		if e.IdempotencyKey == oe.IdempotencyKey {
			gotIt = true
		}
	}
	if !gotIt {
		t.Fatalf("effect not in dispatch list")
	}
	// Claim → second claim loses.
	cl1, _ := c.ClaimEffectDispatch(ctx, r2.RunId, oe.IdempotencyKey, "go-A", 60_000)
	cl2, _ := c.ClaimEffectDispatch(ctx, r2.RunId, oe.IdempotencyKey, "go-B", 60_000)
	if !cl1.Acquired || cl2.Acquired {
		t.Fatalf("CAS lease broken: cl1=%v cl2=%v", cl1.Acquired, cl2.Acquired)
	}
	// Record a dispatch failure that drives the effect to UNKNOWN (no retry).
	if _, err := c.RecordDispatchAttempt(ctx, r2.RunId, oe.IdempotencyKey,
		"simulated lost ack", 0); err != nil {
		t.Fatal(err)
	}
	ge2, _ := c.GetEffect(ctx, r2.RunId, oe.IdempotencyKey)
	if int32(ge2.Effect.Status) != EffectStatusUnknown {
		t.Fatalf("expected UNKNOWN after lost ack; got %v", ge2.Effect.Status)
	}
	// Reconciler observes ABSENT for non-idempotent → FAILED (no re-issue).
	if _, err := c.RecordExternalObservation(ctx, RecordExternalObservationOpts{
		RunID: r2.RunId, Key: oe.IdempotencyKey,
		Resolution: EffectResolutionAbsent,
	}); err != nil {
		t.Fatal(err)
	}
	ge2, _ = c.GetEffect(ctx, r2.RunId, oe.IdempotencyKey)
	if int32(ge2.Effect.Status) != EffectStatusFailed {
		t.Fatalf("NON_IDEMPOTENT + ABSENT must land FAILED; got %v", ge2.Effect.Status)
	}

	// session + event roundtrip
	if _, err := c.CreateSession(ctx, "a", "u", "go-sess", `{"k":1}`); err != nil {
		t.Fatal(err)
	}
	ae, err := c.AppendEvent(ctx, "a", "u", "go-sess",
		&pb.EventRecord{Id: "e1", InvocationId: "inv", Author: "user", ContentJson: "{}", ActionsJson: "{}"},
		`{"k":2}`)
	if err != nil || ae.Event == nil {
		t.Fatalf("append: %v ev=%v", err, ae.Event)
	}
	gs, err := c.GetSession(ctx, "a", "u", "go-sess", 0)
	if err != nil || !gs.Found {
		t.Fatalf("get_session: %v found=%v", err, gs.Found)
	}
	if len(gs.Session.Events) != 1 || gs.Session.StateJson == "" {
		t.Fatalf("session: %#v", gs.Session)
	}
}

File: tape/sdk/typescript/test/client.test.ts
// End-to-end smoke: spawn the Rust tape-server with an in-memory store, then
// round-trip the full effect lifecycle (begin_run -> record_decision ->
// begin_effect/begin_effect-short-circuit -> complete_effect -> get_effect ->
// register_compensation -> set/admit/charge_budget -> end_run).
// Skips if the tape-server binary isn't built.

import test from 'node:test';
import assert from 'node:assert/strict';
import { spawn, type ChildProcess } from 'node:child_process';
import { existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import * as net from 'node:net';
import { TapeClient, RunStatus, EffectStatus, EffectSemantics, EffectDispatchMode, EffectResolution } from '../src/index.ts';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SERVER_BIN = join(__dirname, '..', '..', '..', 'server', 'target', 'debug', 'tape-server');

function freePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const s = net.createServer();
    s.listen(0, '127.0.0.1', () => {
      const p = (s.address() as net.AddressInfo).port;
      s.close(() => resolve(p));
    });
    s.on('error', reject);
  });
}

async function waitFor(host: string, port: number, timeoutMs = 15_000): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const ok = await new Promise<boolean>((r) => {
      const s = net.connect(port, host);
      s.on('connect', () => { s.end(); r(true); });
      s.on('error', () => { r(false); });
    });
    if (ok) return;
    await new Promise((r) => setTimeout(r, 100));
  }
  throw new Error(`server never came up at ${host}:${port}`);
}

let proc: ChildProcess | null = null;
let url = '';

test.before(async () => {
  if (!existsSync(SERVER_BIN)) {
    test.skip(`tape-server not built — run \`cargo build\` in ${SERVER_BIN}/../..`);
    return;
  }
  const port = await freePort();
  proc = spawn(SERVER_BIN, ['--listen', `127.0.0.1:${port}`, '--store', 'memory'], {
    stdio: 'ignore',
    env: { ...process.env, RUST_LOG: 'tape_server=warn' },
  });
  await waitFor('127.0.0.1', port);
  url = `tape://127.0.0.1:${port}`;
});

test.after(() => {
  if (proc) { proc.kill('SIGTERM'); proc = null; }
});

test('TapeClient round-trips a full effect lifecycle', async () => {
  if (!proc) return; // skipped
  const c = new TapeClient(url);
  try {
    const begun: any = await c.beginRun({ appName: 'a', userId: 'u', sessionId: 'ts-smoke',
      invocationId: 'inv-ts', leaseOwner: 'test', leaseTtlMs: 60_000 });
    assert.equal(begun.resumed, false);
    const rid: string = begun.runId;

    await c.recordDecision({ runId: rid, decisionIndex: 0, model: 'm',
      responseJson: '{"plan":1}', rationale: '', policyVersion: 'p1' });
    const got: any = await c.getDecision({ runId: rid, decisionIndex: 0 });
    assert.equal(got.found, true);

    const be: any = await c.beginEffect({ runId: rid, decisionIndex: 0,
      toolName: 'execute_sweep', callIndex: 0, requestJson: '{}' });
    assert.equal(be.status, EffectStatus.PENDING);
    assert.equal(be.idempotencyKey, `${rid}/decision-0/execute_sweep/0`);

    // second begin -> short-circuits to the existing PENDING (the same effect row)
    const be2: any = await c.beginEffect({ runId: rid, decisionIndex: 0,
      toolName: 'execute_sweep', callIndex: 0, requestJson: '{}' });
    assert.equal(be2.idempotencyKey, be.idempotencyKey);
    assert.equal(be2.status, EffectStatus.PENDING);

    await c.completeEffect({ runId: rid, idempotencyKey: be.idempotencyKey,
      status: EffectStatus.CONFIRMED, responseJson: '{"wire_id":"w1"}' });
    const ge: any = await c.getEffect({ runId: rid, idempotencyKey: be.idempotencyKey });
    assert.equal(ge.found, true);
    assert.equal(ge.effect.status, EffectStatus.CONFIRMED);
    assert.match(ge.effect.responseJson, /wire_id/);

    await c.registerCompensation({ runId: rid, effectKey: be.idempotencyKey,
      kind: 'reverse_wire', payloadJson: '{}' });
    const obs: any = await c.listObligations({ runId: rid, onlyUnresolved: true });
    assert.equal(obs.obligations.length, 1);

    // budget admit/charge
    await c.setBudget({ runId: rid, usdCap: 1.0, tokenCap: 0 });
    let adm: any = await c.admitBudget({ runId: rid, usdEstimate: 0.5 });
    assert.equal(adm.admitted, true);
    await c.chargeBudget({ runId: rid, usd: 0.9 });
    adm = await c.admitBudget({ runId: rid, usdEstimate: 0.5 });
    assert.equal(adm.admitted, false);

    // timer roundtrip
    const tr: any = await c.setTimer({ runId: rid, fireAtMs: Date.now() - 1000,
      kind: 'gate_timeout', payloadJson: '{"gate":"g1"}' });
    const due: any = await c.listDueTimers({ claim: true, limit: 50 });
    assert.ok(due.timers.some((t: any) => t.timerId === tr.timerId));

    await c.endRun({ runId: rid, status: RunStatus.TERMINAL });
    const fresh: any = await c.getRun(rid);
    assert.equal(fresh.status, RunStatus.TERMINAL);

    // a re-begin_run finds the existing (TERMINAL) run
    const again: any = await c.beginRun({ appName: 'a', userId: 'u', sessionId: 'ts-smoke',
      invocationId: 'inv-ts', leaseOwner: 'test', leaseTtlMs: 60_000 });
    assert.equal(again.resumed, true);
    assert.equal(again.runId, rid);
    assert.equal(again.status, RunStatus.TERMINAL);
  } finally { c.close(); }
});

test('TapeClient round-trips the outbox / non-idempotent contract', async () => {
  if (!proc) return; // skipped
  const c = new TapeClient(url);
  try {
    const begun: any = await c.beginRun({ appName: 'a', userId: 'u', sessionId: 'ts-outbox',
      invocationId: 'inv-ts-outbox', leaseOwner: 'test', leaseTtlMs: 60_000 });
    const rid: string = begun.runId;

    // Server refuses NON_IDEMPOTENT + INLINE.
    await assert.rejects(c.beginEffect({
      runId: rid, decisionIndex: -1, toolName: 'wire_money',
      semantics: EffectSemantics.NON_IDEMPOTENT,
      dispatchMode: EffectDispatchMode.INLINE,
    }));

    // NON_IDEMPOTENT + OUTBOX with a business key is accepted.
    const oe: any = await c.beginEffect({
      runId: rid, decisionIndex: -1, toolName: 'wire_money',
      requestJson: '{"amount":100}',
      semantics: EffectSemantics.NON_IDEMPOTENT,
      dispatchMode: EffectDispatchMode.OUTBOX,
      businessKey: 'ts:bk-1', connector: 'bank.wire',
    });
    assert.equal(oe.status, EffectStatus.PENDING);

    // Visible to the outbox dispatcher.
    const list: any = await c.listEffectsToDispatch({ connector: 'bank.wire', limit: 50 });
    assert.ok(list.effects.some((e: any) => e.idempotencyKey === oe.idempotencyKey));

    // CAS lease — second claim loses.
    const cl1: any = await c.claimEffectDispatch({ runId: rid, idempotencyKey: oe.idempotencyKey, claimer: 'ts-A' });
    const cl2: any = await c.claimEffectDispatch({ runId: rid, idempotencyKey: oe.idempotencyKey, claimer: 'ts-B' });
    assert.equal(cl1.acquired, true);
    assert.equal(cl2.acquired, false);

    // Lost ack → UNKNOWN.
    await c.recordDispatchAttempt({ runId: rid, idempotencyKey: oe.idempotencyKey,
      error: 'simulated lost ack', nextDispatchAtMs: 0 });
    let ge: any = await c.getEffect({ runId: rid, idempotencyKey: oe.idempotencyKey });
    assert.equal(ge.effect.status, EffectStatus.UNKNOWN);

    // Reconciler observes ABSENT for non-idempotent → FAILED (no re-issue).
    await c.recordExternalObservation({ runId: rid, idempotencyKey: oe.idempotencyKey,
      resolution: EffectResolution.ABSENT });
    ge = await c.getEffect({ runId: rid, idempotencyKey: oe.idempotencyKey });
    assert.equal(ge.effect.status, EffectStatus.FAILED);
  } finally { c.close(); }
});
